readme.md

# Dev Setup
* Clone the repo
* Navigate your Chrome browser to `chrome://extensions/`
* Click `Load Unpacked Extension` button at the top and select monoTab folder
* Make sure it is 'enabled' after loading it

# The Why

**Different from typical user-lists apps**
Reason it _needs to be on a new tab_ is because most user-list apps out there:
Pocket, flipboard, etc tend to become graveyards of "set and forget". When our lists
aren't in front of us they are often are never revisited.

MonoTab is different in that it hijacks the new tab screen in Chrome. A screen
we all see many times each day. The goal is to create something worth of taking over this
prime real estate. It needs to have a remarkable ux — something users will enjoy interacting with.

**Memory Savings**
Save your browser some memory of having all those pages open at once. Write more here.

changelog.md
#CHANGELOG:
### Progress log prior to git setup
---

## Current Goal
* Fix naming conventions

## Next Goal
* Setup github
* Max at 7 lists
* Handle inbox zero
* If was last item on list and opened, or deleted last
* Only show nav if more than two lists
* Only show add list button if more than one link exists

---
# 58 Hours Spent

## Changelog
**Sunday July 17 2016** (9h)
* Fix set count on link open and delete link
* Add no lists state
* Scroll based on mouse down and  mouse x position
* Change list floating behavior on list count change
* Add horizontal scrolling easing and link to nav
* Handle save/sort of new list
* On new list add to nav and click that btn for scroll
* Add funky names helpers

**Wednesday June 15 2016** (1.3h)
* Fix dragula
* Create new list

**Saturday June 11 2016** (1.9h)
* Fix list display
* Create new list beginning

**Friday June 10 2016** (1.5h)
* make _Return_ in top textarea blur/save it
* make _Return_ in content-editable blur/save it
* Update nav buttons on listTitle change
* v1 planning & styles updates

**Wednesday June 8 2016** (2.5h)
* Add list navbar w button per list title
* Add about modal

**Saturday June 4 2016** (2.5h)
* Configure saving list titles
* Configure link counts for multiple lists

**Friday June 3 2016** (1h)
* Configure drag/drop for multiple lists
* Drag drop between lists
* Make color scheme pure mono

**Thursday June 2 2016** (40m)
* Reconfigure deleting links for multiple lists

**Tuesday May 31 2016** (1h)
* Add new lists

**Monday May 30 2016** (5h)
* Change data structure to add to named list
* Fix headerText wordcount per font size

**Tuesday May 17 2016** (.7h)
* Add spinner for list loading
* Shorten deleter in-animation

**Friday May 13 2016** (2h)
* Add filters to favicon

**Wednesday May 11 2016** (4h)
* Switch favicon Library
* Fix scrollability
* Style favicons

**Monday May 9 2016** (1.5h)
* Favicons
* Select all text when text area is entered

**Sunday May 8 2016** (2h)
* Save new positions to chromeStorage after sorting list

**Friday May 6 2016** (5h)
* See if it's easy to hijack cmmd + s (confirmed)
* Install dragula sortable library / style
* Uninstall sortable library and find new one
* Style sortable library and find bug for mouse-cursor
* Add sortable library
* Restructure filesystem

**Tuesday May 3 2016** (6h)
* Add Count to titlebar have it update on .deleter click
* Add listener in url bar click to close tab and save it (Instead of two clicks
  — one to open a popup and another to cut the current tab or cut all tabs)
* Don't add pageload indicators if user cmmd + clicked link (opened in new tab)
* Fix on link click handeler form thwarting the delete handler
* Fix first item disappearing bug on hover-exit
* Attempt to fix bottom margin bug

**Monday May 2 2016** (3h)
* Rework tooltip style
* Add exit animations

**Sunday May 1 2016** (4h)
* Save Items to chrome-storage on url-bar button click
* Show Items in storage on New Tabs page
* Styles beginning

**Saturday Apr 30 2016** (4h)
* Setup file structure and get js running

js/helpers.js
// helping out
var h =  {
  gimmeDateString:  function() {
    var monthNames = [
      "Jan", "Feb", "Mar",
      "Apr", "May", "Jun", "Jul",
      "Aug", "Sep", "Oct",
      "Nov", "Dec"
    ]

    var date = new Date()
    var day = date.getDate()
    var monthIndex = date.getMonth()
    var year = date.getFullYear()

    return day + ' ' + monthNames[monthIndex] + ' ' + year
  },

  rando: function(arr) {
    return arr[Math.floor(Math.random() * arr.length)]
  },

  gimmeListName: function() {
    var firstWord = ['chamfered', 'holstered', 'bandied', 'cacheted', 'sourced',
      'baked-in', 'sequestered', 'bedeviled', 'abdicated', 'goose-stepped',
      'couched', 'vacuum', 'cadging', 'oblique', 'cordoned', 'nerd', 'known',
      'placated', 'crackling', 'hearty', 'sheathed', 'du jour', 'laissez-faire',
      'kludgy', 'effusive', 'febrile', 'elegaic', 'concussive', 'incontrovertable'
    ]

    var secondWord = ['tacos', , 'modicum', 'walrus', 'fanny pack',
      'fingerstache', 'umami', 'rogan', 'mount', 'rights', 'gills',
      'try', 'sound-box', 'laser-burn', 'belief', 'salad', 'disruption',
      'warrior', 'square', 'salvo', 'sprinkle', 'euthymia', 'fusiform',
      'malaprops', 'curlicues', 'carbuncle', 'gimble', 'flogging',
      'peals', 'goldilocks', 'armpits', 'troglodyte', 'cosmic', 'testbed',
      'kerfuffle', 'curio', 'gubbins', 'constituents', 'geronimo',
      'coquetry', 'dilliance', 'sophist', 'chrysalis', 'vuperdu', 'dossier',
      'welter', 'thrum', 'lunacy', 'truncheon', 'pugs'
    ]

    return this.rando(firstWord) + " " + this.rando(secondWord)
  },

  gimmeShoutOut: function() {
    var phrase = [
      "Go skip some stones in a swimming pool.",
      "People are better than links. Go talk to someone.",
      "You destroyer of worlds.",
      "You warrior, you.",
      "Ka-pow, ka-boom. Axe-kick to the dome.",
      "Time for a taco.",
      "Thank you! XOXO, Your Ram."
    ]

    return this.rando(phrase)
  }
}

js/background.js

chrome.browserAction.onClicked.addListener(function() {
  chrome.tabs.query({active: true}, function(data) {
    // Get setup active tab's
    var tabToSave = {
      id: data[0].id,
      url: data[0].url,
      title: data[0].title,
      date: gimmeDateString()
    }
    // Add tab to storage
    chrome.storage.sync.get('monotabdata', function(tabsObj) {
      if (Object.keys(tabsObj).length ==  0) {
        var tmp = {'defaultList': [tabToSave]}
      } else {
        var tmp = JSON.parse(tabsObj.monotabdata)
        if (Object.keys(tmp).length == 0) {
          console.log('tmp was seen as 0')
          tmp = {'defaultList': [tabToSave]}
        } else if (Object.keys(tmp).length == 1 && tmp.titles !== undefined) {
          console.log('tmp was seen as 1 and has titles')
          // Mutate tmp to delete titles and save it as the only tab
          tmp = {'defaultList': [tabToSave]}
        } else {
          // if defaultList already exists we've made it this far. We add to this list
          tmp.defaultList.unshift(tabToSave)
        }
      }

      chrome.storage.sync.set({'monotabdata': JSON.stringify(tmp)})
    })
    // Close the active tab after saving
    chrome.tabs.remove(data[0].id)
  })
})

function gimmeDateString() {
  // Dates!
  var monthNames = [
    "Jan", "Feb", "Mar",
    "Apr", "May", "Jun", "Jul",
    "Aug", "Sep", "Oct",
    "Nov", "Dec"
  ]

  var date = new Date()
  var day = date.getDate()
  var monthIndex = date.getMonth()
  var year = date.getFullYear()

  return day + ' ' + monthNames[monthIndex] + ' ' + year
}

function extend(a, b){
  for(var key in b)
    if(b.hasOwnProperty(key))
      a[key] = b[key];
  return a;
}

js/monoTab.js
// On Ready
$(document).ready(function() {
  /*
      ———————— INIT CALLS ————————
  */
  var drake = dragula() // drag/drop library
  drake.on('dragend', saveSort)
  setupHeaderText()
  setupLinks(drake)

  /*
      ———————— LISTENERS ————————
  */
  $('body').on('click', '.deleter', handleDeleterClick)
  $('body').on('click', '.lists-container a', handleLinkClick)
  $('body').on('click', '.nav-btn', handleNavClick)
  $('body').on('click', '.add-list', handleAddList)
  $('body').on('click', '.scrim', handleScrimClick)
  $('body').on('click', '.hi', handleAboutClick)
  $('body').on('change blur', '#header-text', saveHeaderText)
  $('body').on('keydown', '#header-text', resizeHeaderText)
  $('body').on('change blur', '.title-field', saveListName)
  $('body').on('keydown', '.title-field', checkForReturnKey)
  $(document).mousedown(function() { s.mouseIsDown = true })
    .mouseup(function() { s.mouseIsDown = false })
  $('body').on('mousemove', handleMouseMove)

  function handleAddList() {
    var title = h.gimmeListName()
    var id = Date.now().toString()
    var listHtml = createList(id, title, '', 0)

    $('.lists-container').append(listHtml)

    // Setup the list with drag/drop
    drake.containers.push(document.getElementById(id))

    // Set lists-container width
    handleContainerWidth()

    // Push new list into navigation
    var newNavBtn = createNavBtn([id, title])
    $('.nav-btns').append(newNavBtn)

    // Click the new button to bring it into view
    $('[data-list='+ id +']').click()

    // Show nav if there are more than 3 lists
    showNavAsNeeded()
  }

}) // End On Ready

/*
    ———————— GLOBAL STATE ————————
*/
var s = {
  linkTotal: 0,
  mouseIsDown: false,
  pageWidth: undefined
}

/*
    ———————— FUNCTIONS ————————
*/
function handleMouseMove(e) {
  if (s.mouseIsDown) {
    s.pageWidth ? s.pageWidth : s.pageWidth = $('body').width()

    if (e.pageX / s.pageWidth > .85) {
      startHorzScroll('right')
    } else if (e.pageX / s.pageWidth < .15) {
      startHorzScroll('left')
    } else {
      stopHorzScroll()
    }
  }
}

function startHorzScroll(dir) {
  var dist = dir === 'left'
    ? 0
    : $('.lists-container').width()

  $('.lists-container-housing').stop().animate({
    scrollLeft: dist
  }, 1200, 'easeInSine')
}

function stopHorzScroll() {
  $('.lists-container-housing').stop()
}

function checkForReturnKey(e) {
  if (e.keyCode == 13) {
    e.preventDefault()
    $(this).blur()
    window.getSelection().removeAllRanges()
  }
}

function saveListName() {
  var $this = $(this)
  var listId = $this.attr('name')
  var newTitle = $this.text()
  var btn = $('.nav-btns').find('[data-list="'+listId+'"]')

  // (Optimistically) update nav title
  btn.text(newTitle)
  // Update nav title
  chrome.storage.sync.get('monotabdata', function(tabsArray) {
    var currentTabs = (tabsArray.monotabdata === null) ? {} : JSON.parse(tabsArray.monotabdata)

    if (currentTabs.titles === undefined) {
      currentTabs.titles = {}
      currentTabs.titles[listId] = newTitle
    } else {
      currentTabs.titles[listId] = newTitle
    }

    chrome.storage.sync.set({'monotabdata': JSON.stringify(currentTabs)})
  })

}

function handleDeleterClick(e) {
  e.preventDefault()
  e.stopPropagation()

  var link = $(this).closest('a')
  var linkId = link.data('link-id')
  var listId =  link.closest('.links-list').attr('id')

  deleteLink(linkId, listId)
  link.slideUp(190)
}

function handleLinkClick(e) {
  var link = $(this)
  var linkId = link.data('link-id')
  var listId = link.closest('.links-list').attr('id')
  var animationTime = 190

  if (e.metaKey && e.shiftKey || (e.ctrlKey && e.shiftKey)) {
    return true
  }

  deleteLink(linkId, listId)
  link.slideUp(animationTime)

  if (e.metaKey || e.ctrlKey) {
    return true
  } else {
    $('.loader').fadeIn(animationTime+170)
    $('.load-bar').addClass('gotime')
  }

}

function handleNavClick() {
  var list = $(this).data('list')
  $('.lists-container-housing').stop().animate({
		scrollLeft: $('#' + list).offset().left
	}, 920, 'easeOutQuad')
}

function handleScrimClick() {
  $(this).fadeOut(225)
}

function handleAboutClick() {
  $('.scrim').fadeIn(225)
}

function deleteLink(linkId, listId) {

  var list = $('#' + listId)
  var listCountEl = list.closest('.list').find('.count')
  var newCount = list.find('.link').length - 1

  // decrement state object
  s.linkTotal -= 1
  showAddBtnAsNeeded()

  setCount(listCountEl, newCount)

  chrome.storage.sync.get('monotabdata', function(tabsArray) {
    var currentTabs = (tabsArray.monotabdata === null) ? {} : JSON.parse(tabsArray.monotabdata)
    var listInQuestion = currentTabs[listId]

    var groomedList = listInQuestion.filter(function(el) {
      return el.id !== linkId
    })

    currentTabs[listId] = groomedList

    if (currentTabs[listId].length === 0) delete currentTabs[listId]

    chrome.storage.sync.set({'monotabdata': JSON.stringify(currentTabs)})
  })
}

function saveSort(el) {
  chrome.storage.sync.get('monotabdata', function(tabsArray) {
    var fullSet = (tabsArray.monotabdata === null) ? {} : JSON.parse(tabsArray.monotabdata)

    $.each($('.links-list'), function() {
      var $this = $(this)
      var listId = $this.attr('id')

      var sortedLinks = []
      $this.find('.link').each(function() {
        var $this = $(this)
        var linkObj = {
          id: $this.data('link-id'),
          url:  $this.attr('href'),
          title: $this.data('title'),
          date: $this.data('date')
        }
        sortedLinks.push(linkObj)
      })

      fullSet[listId] = sortedLinks

      // Remove list and associated nav-btn if it is empty after sorting
      if (fullSet[listId].length === 0) {
        var selector = '#' + listId
        $(selector).closest('.list').remove()
        delete fullSet[listId]
        $('[data-list=' + listId + ']').remove()
      }

      // Save title if we have a new list
      if (!fullSet.titles) {
        fullSet.titles = {}
      }
      if (fullSet.titles[listId] == undefined) {
        fullSet.titles[listId] = $(this).closest('.list').find('.title-field').text()
      }
    })


    // Set lists-container width
    handleContainerWidth()

    chrome.storage.sync.set({'monotabdata': JSON.stringify(fullSet)})

    updateCountsOnSort()

    showNavAsNeeded()
  })
}

// Only show nav if there are three or more lists
function showNavAsNeeded() {
  if ($('.nav-btn').length > 2) {
    $('.nav-btns').show()
  } else {
    $('.nav-btns').hide()
  }
}

// Only show add-list button if there is more than one link in state
function showAddBtnAsNeeded() {
  if (s.linkTotal > 1) {
    $('.add-list').show()
  } else {
    $('.add-list').hide()
  }
}

// Used for testing
function whatIsStateNow() {
  chrome.storage.sync.get('monotabdata', function(tabsObj) {
    console.info(tabsObj)
  })
}

// TODO: Make onboarding prettier. Maybe step through slider.
function checkForEmptyState(lists) {
  var listsCount = Object.keys(lists).length

  if (listsCount <= 1 ) {
    if (listsCount === 1 && lists.titles == undefined) {
      return false
    } else {
      $('.spinner').hide()
      $('.empty-state').show()
    }
  }
}

function setupLinks(drake) {
  chrome.storage.sync.get('monotabdata', function(tabsObj) {
    var lists = (Object.keys(tabsObj).length ===  0) ? {} : JSON.parse(tabsObj.monotabdata)

    checkForEmptyState(lists)

    var listInfoForNav = []
    // For each list in lists
    $.each(lists, function(x, list) {
      var listId = x
      var listCount = list.length
      var linksObjs = []

      if (listId !== 'titles') {
        $.each(list, function(y, linkObj) {
          linksObjs.push(linkObj)
          // update links total in state object
          s.linkTotal += 1
        })

        // Map over our array of link objects and turn them into html
        var linkHtmlArray = linksObjs.map(createListItems)
        var linksHtml = linkHtmlArray.join('')
        // See if a user-saved title exists, if so pass it in for createList
        var listTitle = lists.titles === undefined
            ? listId
            : lists.titles[listId]

        // Setup nav with list titles
        listInfoForNav.push([listId, listTitle])

        // Create list HTML
        var listHtml = createList(listId, listTitle, linksHtml, listCount)
        $('.lists-container').append(listHtml)

        // Setup the list with drag/drop
        drake.containers.push(document.getElementById(listId))
      }
    })

    // Set lists-container width
    handleContainerWidth()

    // Add navigation buttons to navbar
    var navBtns = listInfoForNav.map(createNavBtn)
    var navHolster = $('.nav-btns')

    navHolster.append(navBtns)

    // If we have less than 3 lists don't show nav buttons
    showNavAsNeeded()

    // If we have only one link item we hide the add-list button
    showAddBtnAsNeeded()

    $('.spinner').hide()
  })
}

function createNavBtn(array){
  var listId = array[0]
  var listName = array[1]
  return '<button type="button" class="nav-btn" data-list="'+listId+'">'+listName+'</button>'
}

function createList(id, title, linksItems, count) {
  var list =
    '<div class="list">'+
      '<div class="list-title">'+
        '<span class="title-field" contenteditable="true" name="'+id+'">'+title+'</span>'+
        '<div class="goofwrap">&nbsp;('+
          '<span class="count-holster">'+
            '<span class="count">'+ count +'</span>'+
          '</span>'+
        ')</div>'+
      '</div>'+
      '<div class="links-list" id="'+ id +'">'+
        linksItems +
      '</div>'+
    '</div>'

  return list
}

function createListItems(link) {
  if (link.title == "MonoTab") {
    return
  }

  var prettyUrl = link.url.replace(/^(https?:\/\/)?(www\.)?/,'').replace(/\/$/, "")
  var prettyUrlTrunced = prettyUrl.trunc(47)
  var baseUrl = prettyUrl.split('/')[0]
  var titleTruced = link.title.trunc(47) || 'No Title'
  var date = link.date || 'Date Unknown'
  var favicon = '<img width="16" height="16" class="favicon" src="http://www.google.com/s2/favicons?domain=' + baseUrl + '" />'

  var linkUI =
    '<a class="link" data-link-id="' +link.id+'" href="' +link.url+ '" data-date="' +date+ '" data-title="' +link.title+ '">' +
    favicon +
    titleTruced +
      '<div class="tooltip">' +
         '<div class="tooltip-item-wrap">' +
          '<div class="tooltip-title mid">URL</div>' +
          '<div class="tooltip-data mid">' +prettyUrlTrunced+ '</div>' +
        '</div>' +
         '<div class="tooltip-item-wrap">' +
          '<div class="tooltip-title">ADDED</div>' +
          '<div class="tooltip-data">' +date+ '</div>' +
        '</div>' +
      '</div>' +
      '<div class="deleter">&times</div>' +
    '</a>'

  return linkUI
}

function updateCountsOnSort() {
  $.each($('.count'), function() {
    var $currentCount = $(this)
    var currentValue = parseInt($currentCount.text())
    var newCount = $currentCount.closest('.list').find('.link').length

    if (currentValue !== newCount) setCount($currentCount, newCount)
  })
}

function setCount(countToChange, newCount) {
  countToChange.addClass('exit-up')

  setTimeout(function() {
    var holster = countToChange.closest('.count-holster')
    countToChange.hide().removeClass('count')
    holster.append(
      '<span class="count below">' +newCount+ '</span>'
    )
    setTimeout(function() {
      holster.find('span').removeClass('below')
    }, 17)
  }, 200)
}

function saveHeaderText() {
  var newMainText = $('#header-text').val()

  if (!newMainText) return

  resizeHeaderText()

  chrome.storage.sync.set({'mainText': newMainText})
}

function setupHeaderText() {
  chrome.storage.sync.get('mainText', function(text) {
    var setupText = text.mainText || "MonoTab"

    $('#header-text').val(setupText)
    resizeHeaderText()
  })
}

// Resize font-size based on how may words are within the message
function resizeHeaderText(e) {
  var headerText = $('#header-text')
  if (e !== undefined) {
    if (e.keyCode == 13) {
      e.preventDefault();
      headerText.blur()
      return false
    }
  }

  var numWords = headerText.val().split(" ").length

  if ((numWords >= 1) && (numWords < 3)) {
    headerText.css("font-size", "120px")
  }
  else if ((numWords >= 3) && (numWords < 5)) {
    headerText.css("font-size", "90px")
  }
  else if ((numWords >= 3) && (numWords < 10)) {
    headerText.css("font-size", "70px")
  }
  else if ((numWords >= 10) && (numWords < 15)) {
    headerText.css("font-size", "50px")
  }
  else if ((numWords >= 20) && (numWords < 27)) {
    headerText.css("font-size", "40px")
  }
  else if ((numWords >= 30) && (numWords < 40)) {
    headerText.css("font-size", "33px")
  }
  else {
    headerText.css("font-size", "20px")
  }
}

function handleContainerWidth() {
  var listsCount = $('.list').length

  if (listsCount > 2) {
    $('.list').css('float', 'left')

    var newWidth = (listsCount - 2) * 530 + 1200
    $('.lists-container').css('width', newWidth + 'px')
  } else {
    $('.list').css('float', 'none')
    $('.lists-container').css('width', '100%')
  }
}

// Decorate native String obejct with trucation
String.prototype.trunc = String.prototype.trunc ||
  function(n) {
    return (this.length > n) ? this.substr(0, n - 1) + '<span class="lip">&hellip;</span>' : this
  }

// function imageExists(src, callback) {
//   var img = new Image()
//   img.onload = function() {
//     callback(src, true)
//     state.hasFavicon = true
//   }
//   img.onerror = function() {
//     callback(src, false)
//     state.hasFavicon = false
//   }
//   // img.onload = function() {
//   //   callback(src, true)
//   //   fImg = true
//   // }
//   // img.onerror = function() {
//   //   callback(src, false)

//   // }
//   img.src=src
// }

// Make favicon loading more robust
// Work out this conditional

// var baseUrl = 'reactforbeginners.com';

// imageExists('http://' + baseUrl +'/favicon.ico', function(src, exists) {
//     if (exists) {
//         console.log('worked');
//         $('<img>').attr('src', src).appendTo('body');
//     } else {
//         console.log('failed');
//         $('<img>').attr('src', "http://www.google.com/s2/favicons?domain=" + baseUrl).appendTo('body');
//     }
// });


// function imageExists(src, callback) {
//     var img = new Image();
//     img.onload = function() {
//         callback(src, true);
//     };
//     img.onerror = function() {
//         callback(src, false);
//     };
//     img.src=src;
// }




//or


// works but dont' want to use
  // faviconUrl = 'http://favicon.yandex.net/favicon/' + baseUrl